add selective tracing support to SpawnerTraceExt

kat-perez opened the pull request:
Add spawn_with_trace(), spawn_no_trace() and named variants to provide fine-grained control over task tracing. Supports both default tracing (rtos-trace) and selective tracing (rtos-trace-selective) modes while maintaining backward compatibility.

kat-perez committed:
add selective tracing support to SpawnerTraceExt

Add spawn_with_trace(), spawn_no_trace() and named variants to provide
fine-grained control over task tracing. Supports both default tracing
(rtos-trace) and selective tracing (rtos-trace-selective) modes while
maintaining backward compatibility.

lulf commented on embassy-executor/src/raw/mod.rs:
```diff
@@ -94,6 +94,8 @@ pub(crate) struct TaskHeader {
     #[cfg(feature = "trace")]
     pub(crate) id: u32,
     #[cfg(feature = "trace")]
+    pub(crate) trace_excluded: bool,
```
This should probably be guarded by a trace-mode-selective flag to save some space for programs not using it.

kat-perez replied:
@lulf (bump)

fg-cfh replied:
Nit: I propose we call this trace_enabled and reverse the logic. Positive logic is more ergonomic.

kat-perez committed:
add trace-mode-selective flag

- update rtos-trace selective to depend on trace-mode-selective
- gate trace_excluded field behind trace-mode-selective instead of trace
- reduce memory overhead when selective tracing is not needed

kat-perez commented:
@lulf the ci/build job seemed to fail to error 137. Could I get a re-run for this job?

lulf commented:
bender run

fg-cfh commented on embassy-executor/Cargo.toml:
```diff
@@ -95,8 +95,12 @@ executor-thread = []
 trace = []
-## Enable support for rtos-trace framework
+## Enable selective tracing support
+trace-mode-selective = ["trace"]
```
I propose that this feature flag be orthogonal to the trace and rtos-trace features, i.e. you do --features "trace trace-selective" and --features "rtos-trace trace-selective". Generating the cross product of features doesn't scale well.

You could implement a static assertion that warns if "trace-selective" is used without selecting a trace backend, too.

This change will also simplify some of your cfg logic below.

fg-cfh commented on embassy-executor/src/raw/trace.rs:
```diff
@@ -181,6 +181,12 @@ pub trait TaskRefTrace {
     fn set_id(&self, id: u32);
+
+    /// Check if a task is excluded from tracing
+    fn is_trace_excluded(&self) -> bool;
```
Either remove the is_ prefix in the method or add it to the variable, too ([C-Getter](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter))

fg-cfh commented on embassy-executor/src/raw/trace.rs:
```diff
@@ -205,6 +211,49 @@ impl TaskRefTrace for TaskRef {
+/// - `rtos-trace-selective`: Only trace tasks that are explicitly enabled for tracing
+/// - Default `rtos-trace`: Trace all tasks by default
+#[inline]
+fn should_trace_task(task: &TaskRef) -> bool {
```
I don't like the distinction between rtos-trace and embassy trace on the API level. IMO this is a leaky abstraction.

If the distinction should be held up then at least this rtos-trace specific function should be named as such: should_rtos_trace_task().

fg-cfh commented on embassy-executor/src/spawner.rs:
```diff
@@ -196,6 +201,50 @@ pub trait SpawnerTraceExt {
+    ///
+    /// # Returns
+    /// Result indicating whether the spawn was successful
+    fn spawn_named_with_trace<S>(&self, name: &'static str, token: SpawnToken<S>) -> Result<(), SpawnError>;
```
Same argument as for the feature flags: Cross product of features doesn't scale. See @Dirbaio 's argument on Matrix, too. I think we need a different approach here, e.g. task metadata as proposed by Dario.

fg-cfh commented on embassy-executor/src/spawner.rs:
```diff
@@ -212,6 +261,87 @@ impl SpawnerTraceExt for Spawner {
+                #[cfg(feature = "rtos-trace-selective")]
+                task.set_trace_excluded(true);
+                #[cfg(not(feature = "rtos-trace-selective"))]
+                task.set_trace_excluded(false);
```
I propose that methods related to this feature should not be present at all when not needed (i.e. hidden behind the same feature flag, too).

fg-cfh commented on embassy-executor/src/raw/trace.rs:
```diff
@@ -384,14 +439,17 @@ where
-            };
-            rtos_trace::trace::task_send_info(task.id(), info);
+            // Only send task info for tasks that should be traced
+            if should_trace_task(&task) {
```
This method should not be generated when the feature is not present.

fg-cfh reviewed (commented):
I'm new to this community, so I won't block the PR but according to my personal standards it is not ready to be merged:
- The newly introduced feature should work for all kinds of tracing (i.e. embassy _and_ rtos_trace) alike.
- The feature flags are confusing and unnecessarily complex/verbose. Can and should be simplified.
- Runtime and config-time logic is mixed up. Runtime logic should not be constant. If something becomes contant then it should be made constant by config, not by having runtime methods return constant values IMHO.

Dirbaio commented:
Hello,

The pull request backlog on the Embassy repo has grown uncontrollably. In an effort to clean it up, we're declaring "pull request bankruptcy" and closing all pull requests older than 1 year.

We've added a new [review queue](https://github.com/embassy-rs/embassy/blob/main/CONTRIBUTING.md) that works in a FIFO way, to avoid pull request reviews stalling for so long in the future.

If you're still interested in contributing this PR, you're welcome to open a new one. Please rebase on top of latest main branch if you do so the changes merge cleanly.

Thank you for your understanding!
